### DISCUSSION TITLE
Use ipcRenderer from components, compilation fails

### DISCUSSION THREAD

**Discussion by szekelyisz** [quasarframework/quasar#10410] [CLI - Electron mode] [answered]
Using `ipcRenderer` from a component method directly or through an import makes compilation fail.

Component.vue

```typescript
[...]
import { ipcRenderer } from 'electron';

export default class Component extends Vue {
  sendMessage() {
    ipcRenderer.send('send-message', this.message);
  }
}
```

Error message:

```
 App •  DONE  • "Renderer" compiled with errors • 6977ms



 App •  ERROR  •  Renderer  in ./node_modules/electron/index.js

Module not found: Can't resolve imported dependency "fs"
Did you forget to install it? You can run: yarn add fs

 App •  ERROR  •  Renderer  in ./node_modules/electron/index.js

Module not found: Can't resolve imported dependency "path"
Did you forget to install it? You can run: yarn add path


 App •  COMPILATION FAILED  • Please check the log above for details.
```

It looks like the renderer is trying to call node APIs.

My `quasar info`:

```
Operating System - Linux(5.11.0-31-lowlatency) - linux/x64
NodeJs - 12.22.5

Global packages
  NPM - 6.14.14
  yarn - 1.22.5
  @quasar/cli - 1.2.1
  @quasar/icongenie - Not installed
  cordova - Not installed

Important local packages
  quasar - 2.0.3 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app - 3.0.4 -- Quasar Framework local CLI
  @quasar/extras - 1.10.10 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.1.5 -- The progressive JavaScript framework for buiding modern web UI.
  vue-router - 4.0.10
  vuex - Not installed
  electron - 13.1.9 -- Build cross platform desktop apps with JavaScript, HTML, and CSS
  electron-packager - Not installed
  electron-builder - Not installed
  @babel/core - 7.15.0 -- Babel compiler core.
  webpack - 5.50.0 -- Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jsx, es7, css, less, ... and your custom stuff.
  webpack-dev-server - 4.0.0-beta.3 -- Serves a webpack app. Updates the browser on changes.
  workbox-webpack-plugin - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  typescript - 4.2.2 -- TypeScript is a language for application scale JavaScript development
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*
```

Is this a bug or do I need to inject the entire `fs` and `path` modules? That doesn't seem right.

**Comment by ontwikkelfabriek** [comment, marked as answer]
@szekelyisz 

First , I am not an Electron expert.  But I will try:

You cannot call `ipcRenderer` directly from the **UI part** of an Electron app ( .vue files ). You have to use a **bridge** to talk to the **nodejs part** of an Electron app. See:
https://quasar.dev/quasar-cli/developing-electron-apps/electron-preload-script#introduction

**Reply by szekelyisz** [reply]
Thanks for your answer. I read about it more and it seems like a new electron security feature laced with webpack breaking changes (no automatic polyfills from v5 on).

I ended up using [electron-rpc-api](https://github.com/vladimiry/electron-rpc-api) for this purpose because it can do it with type safety. It took a while to figure out how to use it from Quasar though. Maybe I'll make an extension for it if it seems super useful.

**Comment by fprijate** [comment]
The simplest solution is  [VueUse](https://vueuse.org/electron/readme.html)

INSTALL
 npm install @vueuse/core
 npm install @vueuse/electron

USAGE:

import { useIpcRenderer} from "@vueuse/electron";
const ipcRenderer = useIpcRenderer();

// you can nov use ipcRenderer as usual
const port =ipcRenderer.sendSync('PORT');

**Comment by pikil** [comment]
Quasar has got the process for that already - both secure and straightforward.
Here is the easiest way (no other plugins involvement!, in addition to reply from @ontwikkelfabriek ):

src-electron/electron-preload.js
```
import { contextBridge, ipcRenderer } from 'electron'

contextBridge.exposeInMainWorld('electronApi', {
  sendMessage: (msg) => {
    ipcRenderer.send('send-message', msg)
  }
})
```

src-electron/electron-main.js
```
import { ipcMain } from 'electron'

ipcMain.on('send-message', (/* event */, msg) => {
    // Do something here...
    // event.returnValue = 'success' // If used sendSync() instead of send()
})
```

Component.vue
```
window.electronApi.sendMessage(message)
```

**Reply by fprijate** [reply]
Yes it works. With VueUse you change import and use the same api cals. In your case you must remove import statements and change api cals.
